drivers/dai/intel/dmic/dmic.c

```diff
@@ -0,0 +1,673 @@
+		dai_dmic_update_bits(
+				dmic,
+				OUTCONTROL0,
+				OUTCONTROL0_FINIT_BIT | OUTCONTROL0_SIP_BIT,
+				OUTCONTROL0_SIP_BIT);
+		break;
+	case 1:
+		LOG_INF("dmic_start(), dmic->fifo_b");
+		/*  Clear FIFO B initialize, Enable interrupts to DSP,
+		 *  Start FIFO B packer.
+		 */
+		dai_dmic_update_bits(dmic, OUTCONTROL1,
+				     OUTCONTROL1_FINIT_BIT | OUTCONTROL1_SIP_BIT,
+				     OUTCONTROL1_SIP_BIT);
+	}
+
+	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++) {
+		mic_a = dmic->enable[i] & 1;
+		mic_b = (dmic->enable[i] & 2) >> 1;
+		fir_a = (dmic->enable[i] > 0) ? 1 : 0;
+		fir_b = (dmic->enable[i] > 0) ? 1 : 0;
+		LOG_INF("dmic_start(), pdm%d mic_a = %u, mic_b = %u", i, mic_a, mic_b);
+
+		/* If both microphones are needed start them simultaneously
+		 * to start them in sync. The reset may be cleared for another
+		 * FIFO already. If only one mic, start them independently.
+		 * This makes sure we do not clear start/en for another DAI.
+		 */
+		if (mic_a && mic_b) {
+			dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
+					     CIC_CONTROL_CIC_START_A_BIT |
+					     CIC_CONTROL_CIC_START_B_BIT,
+					     CIC_CONTROL_CIC_START_A(1) |
+					     CIC_CONTROL_CIC_START_B(1));
+			dai_dmic_update_bits(dmic, base[i] + MIC_CONTROL,
+					     MIC_CONTROL_PDM_EN_A_BIT |
+					     MIC_CONTROL_PDM_EN_B_BIT,
+					     MIC_CONTROL_PDM_EN_A(1) |
+					     MIC_CONTROL_PDM_EN_B(1));
+		} else if (mic_a) {
+			dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
+					     CIC_CONTROL_CIC_START_A_BIT,
+					     CIC_CONTROL_CIC_START_A(1));
+			dai_dmic_update_bits(dmic, base[i] + MIC_CONTROL,
+					     MIC_CONTROL_PDM_EN_A_BIT,
+					     MIC_CONTROL_PDM_EN_A(1));
+		} else if (mic_b) {
+			dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
+					     CIC_CONTROL_CIC_START_B_BIT,
+					     CIC_CONTROL_CIC_START_B(1));
+			dai_dmic_update_bits(dmic, base[i] + MIC_CONTROL,
+					     MIC_CONTROL_PDM_EN_B_BIT,
+					     MIC_CONTROL_PDM_EN_B(1));
+		}
+
+		switch (dmic->dai_config_params.dai_index) {
+		case 0:
+			dai_dmic_update_bits(dmic, base[i] + FIR_CONTROL_A,
+					     FIR_CONTROL_A_START_BIT,
+					     FIR_CONTROL_A_START(fir_a));
+			break;
+		case 1:
+			dai_dmic_update_bits(dmic, base[i] + FIR_CONTROL_B,
+					     FIR_CONTROL_B_START_BIT,
+					     FIR_CONTROL_B_START(fir_b));
+			break;
+		}
+	}
+
+	/* Clear soft reset for all/used PDM controllers. This should
+	 * start capture in sync.
+	 */
+	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++) {
+		dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
+				     CIC_CONTROL_SOFT_RESET_BIT, 0);
+	}
+
+	/* Set bit dai->index */
+	dai_dmic_global.active_fifos_mask |= BIT(dmic->dai_config_params.dai_index);
+	dai_dmic_global.pause_mask &= ~BIT(dmic->dai_config_params.dai_index);
+
+	dmic->state = DAI_STATE_RUNNING;
+	k_spin_unlock(&dmic->lock, key);
+
+	dmic_sync_trigger(dmic);
+
+	LOG_INF("dmic_start(), dmic_active_fifos_mask = 0x%x",
+			dai_dmic_global.active_fifos_mask);
+}
```

Comment:
The dmic_start() starts the stream with CIC and FIR muted as result from blob parse and init, also dmic_stop() sets again the mute bits. This PR does not contain dmic_work() equivalent  so I assume the resulting capture with this driver is zeros? @lyakh has a new solution for driver in SOF without separate task to schedule that would be useful here.

The unmute sequence in dmic_work() ensures the muted CIC feeds in zeros to fill the FIR delay line with zeros. There's no HW reset for RAM so such is needed. That ensures the random PCM samples from FIR 256 samples (usually @ 96 - 192 kHz) delay line can't pass to user space in capture start. The CIC was unmuted at 1ms and FIR at 2ms (times are not critical). Plus there's the log gain ramp with 1ms rate updated HW gain registers.

Reply:
Yes i'm aware that the driver now can pass silence only. As I mentioned, it is not fully tested, mostly because of sof dmic issue that is blocking testing

>@lyakh has a new solution for driver in SOF without separate task to schedule that would be useful here.
Absolutely, it looks very promising. 
Another sollution is to start a 1ms timer and increase volume in async manner